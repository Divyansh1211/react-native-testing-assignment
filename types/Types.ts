export type CriteriaKeys =
  | "length"
  | "uppercase"
  | "lowercase"
  | "numbers"
  | "specialChars"
  | "noRepeatedChars"
  | "noCommonPatterns";

export type StrengthCriteria = {
  [key in CriteriaKeys]: boolean;
};

export interface PasswordStrength {
  level: "Weak" | "Medium" | "Strong";
  score: number;
  maxScore: number;
  criteria: StrengthCriteria;
}

export interface PasswordStrengthValidatorProps {
  onStrengthChange?: (strength: PasswordStrength) => void;
  minLength?: number;
  requireUppercase?: boolean;
  requireLowercase?: boolean;
  requireNumbers?: boolean;
  requireSpecialChars?: boolean;
  preventRepeatedChars?: boolean;
  preventCommonPatterns?: boolean;
}