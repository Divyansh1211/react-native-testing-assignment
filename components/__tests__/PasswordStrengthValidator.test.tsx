import React from "react";
import { fireEvent, render } from "@testing-library/react-native";
import PasswordStrengthValidator from "../PassWordStrengthValidator";
import { PasswordStrength } from "../../types/Types";

const renderWithProps = (props = {}) => {
  const mockFn = jest.fn();
  const utils = render(
    <PasswordStrengthValidator onStrengthChange={mockFn} {...props} />
  );
  return { ...utils, mockFn };
};

describe("PasswordStrengthValidator - Full Suite", () => {
  it("matches snapshot", () => {
    const tree = renderWithProps().toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("initial state is weak", () => {
    const { getByText } = renderWithProps();
    expect(getByText("Strength: Weak")).toBeTruthy();
  });

  it("updates strength when password changes", () => {
    const { getByPlaceholderText, getByText } = renderWithProps();
    fireEvent.changeText(getByPlaceholderText("Enter password"), "Test123!");
    expect(getByText(/Strength:/)).toBeTruthy();
  });

  it("calls onStrengthChange with correct values", () => {
    const { getByPlaceholderText, mockFn } = renderWithProps();
    fireEvent.changeText(getByPlaceholderText("Enter password"), "Test123!");

    expect(mockFn).toHaveBeenCalled();
    const result: PasswordStrength = mockFn.mock.calls[1][0];
    expect(result).toHaveProperty("level");
    expect(result).toHaveProperty("score");
  });

  it("displays correct bar width and color for strong password", () => {
    const { getByPlaceholderText, getByTestId, getByText } = renderWithProps();
    fireEvent.changeText(getByPlaceholderText("Enter password"), "Abc123!@#");

    const bar = getByTestId("strength-bar-fill");
    const strengthLabel = getByText("Strength: Strong");

    expect(strengthLabel).toBeTruthy();
    expect(bar.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          backgroundColor: expect.any(String),
          width: expect.stringMatching(/%/),
        }),
      ])
    );
  });

  it("does not require uppercase when requireUppercase = false", () => {
    const { getByPlaceholderText, mockFn } = renderWithProps({
      requireUppercase: false,
    });

    fireEvent.changeText(
      getByPlaceholderText("Enter password"),
      "lowercase123!"
    );

    expect(mockFn).toHaveBeenCalled(); // ðŸ’¡ Make sure it actually got called

    const result: PasswordStrength = mockFn.mock.calls[1][0];

    expect(result.criteria.uppercase).toBe(true); // âœ… Satisfied because we turned off the requirement
    expect(result.score).toBeGreaterThanOrEqual(1); // âœ… Should have at least some score
  });

  it("handles empty password", () => {
    const { getByPlaceholderText, mockFn } = renderWithProps();
    fireEvent.changeText(getByPlaceholderText("Enter password"), "");

    const result: PasswordStrength = mockFn.mock.calls[0][0];
    expect(result.level).toBe("Weak");
    expect(result.score).toBe(0);
  });

  it("handles very long strong password", () => {
    const longPassword = "A1!a".repeat(30);
    const { getByPlaceholderText, mockFn } = renderWithProps();

    fireEvent.changeText(getByPlaceholderText("Enter password"), longPassword);
    const result: PasswordStrength = mockFn.mock.calls[1][0];

    expect(result.level).toBe("Strong");
    expect(result.score).toBe(result.maxScore);
  });

  it("handles passwords with emojis", () => {
    const password = "Abc123ðŸ˜ŠðŸ˜ŠðŸ˜ŠðŸ˜Š";
    const { getByPlaceholderText, mockFn } = renderWithProps();

    fireEvent.changeText(getByPlaceholderText("Enter password"), password);
    const result: PasswordStrength = mockFn.mock.calls[1][0];

    expect(result.criteria.specialChars).toBe(true);
    expect(result.criteria.noRepeatedChars).toBe(false); // emojis are repeated
  });

  it("renders UI at each strength level", () => {
    const { getByPlaceholderText, getByText } = renderWithProps();
    const input = getByPlaceholderText("Enter password");

    const samples = {
      weak: "abc",
      medium: "Abc123",
      strong: "Abc123!@#",
    };

    fireEvent.changeText(input, samples.weak);
    expect(getByText("Strength: Weak")).toBeTruthy();

    fireEvent.changeText(input, samples.medium);
    expect(getByText("Strength: Medium")).toBeTruthy();

    fireEvent.changeText(input, samples.strong);
    expect(getByText("Strength: Strong")).toBeTruthy();
  });

  it("resets state properly if password is cleared", () => {
    const { getByPlaceholderText, getByText, mockFn } = renderWithProps();
    const input = getByPlaceholderText("Enter password");

    fireEvent.changeText(input, "Abc123!@#");
    fireEvent.changeText(input, "");

    expect(getByText("Strength: Weak")).toBeTruthy();
    const result: PasswordStrength =
      mockFn.mock.calls[mockFn.mock.calls.length - 1][0];
    expect(result.score).toBe(0);
  });
});
