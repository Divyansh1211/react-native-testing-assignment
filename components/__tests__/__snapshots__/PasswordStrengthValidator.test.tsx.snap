// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordStrengthValidator - Full Suite matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "borderRadius": 10,
      "margin": 16,
      "padding": 20,
    }
  }
>
  <Text
    style={
      {
        "color": "black",
        "fontSize": 18,
        "fontWeight": "bold",
        "marginBottom": 10,
      }
    }
  >
    Password
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter password"
    placeholderTextColor="#aaa"
    secureTextEntry={true}
    style={
      {
        "backgroundColor": "white",
        "borderColor": "grey",
        "borderRadius": 4,
        "borderWidth": 1,
        "color": "black",
        "marginBottom": 10,
        "padding": 10,
      }
    }
    value=""
  />
  <View
    style={
      {
        "marginVertical": 10,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#ddd",
          "borderRadius": 4,
          "height": 8,
          "overflow": "hidden",
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 4,
              "height": 8,
            },
            {
              "backgroundColor": "#e74c3c",
              "width": "0%",
            },
          ]
        }
        testID="strength-bar-fill"
      />
    </View>
    <Text
      style={
        [
          {
            "fontWeight": "bold",
            "marginTop": 4,
          },
          {
            "color": "#e74c3c",
          },
        ]
      }
    >
      Strength: 
      Weak
    </Text>
  </View>
  <Text
    style={
      {
        "color": "#B7B7B7",
        "fontSize": 13,
      }
    }
  >
    . Please enter a password
  </Text>
  <View
    style={
      {
        "marginVertical": 10,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 14,
            "marginBottom": 2,
          },
          {
            "color": "red",
          },
        ]
      }
    >
      ✗
       
      Minimum length (8)
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
            "marginBottom": 2,
          },
          {
            "color": "red",
          },
        ]
      }
    >
      ✗
       
      Contains uppercase
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
            "marginBottom": 2,
          },
          {
            "color": "red",
          },
        ]
      }
    >
      ✗
       
      Contains lowercase
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
            "marginBottom": 2,
          },
          {
            "color": "red",
          },
        ]
      }
    >
      ✗
       
      Contains numbers
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
            "marginBottom": 2,
          },
          {
            "color": "red",
          },
        ]
      }
    >
      ✗
       
      Contains special characters
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
            "marginBottom": 2,
          },
          {
            "color": "red",
          },
        ]
      }
    >
      ✗
       
      No repeated characters
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
            "marginBottom": 2,
          },
          {
            "color": "red",
          },
        ]
      }
    >
      ✗
       
      No common patterns
    </Text>
  </View>
</View>
`;
